src/dex.hs

```diff
@@ -81,6 +85,19 @@ replLoop prompt opts = do
                  _ -> return ()
   liftIO $ putStrLn $ pprint result
 
+dexCompletions :: CompletionFunc (StateT TopEnv IO)
+dexCompletions (line, _) = do
+  env <- get
+  let varNames = map pprint $ envNames env
+  -- note: line and thus word and rest have character order reversed
+  let (word, rest) = break (== ' ') line
+  let anywhereKeywords = ["def", "for", "rof", "case", "data", "where", "of", "if", "then", "else", "interface", "instance", "do", "view"]
+  let startoflineKeywords = ["%bench \"", ":p", ":t", ":html", ":export"]
+  let candidates = varNames ++ anywhereKeywords ++
+                   if null rest then startoflineKeywords else []
```

Comment:
This is quite minor, but `++` has complexity proportional to the size of the lhs, so it is usually better to sort the concatenated lists according to increasing length. This concatenates them in the complete reverse order, which is the worst case scenario.

Reply:
TIL, it's an actual linked list i guess

Reply:
Yup, Haskell lists are linked lists. If you want a contiguous arrays you have to use `Data.Vector` (which is a packed array of _boxed values_; if you want to go unboxed there are `Data.Vector.Unboxed` types).